midp/gfx.js

```diff
@@ -793,177 +753,182 @@ var currentlyFocusedTextEditor;
+        this.translate(-this.transX, -this.transY);
+        this.setClip(x1, y1, x2 - x1, y2 - y1, 0, 0, this.contextInfo.context.canvas.width, this.contextInfo.context.canvas.height);
+    }
 
-    Native["javax/microedition/lcdui/Graphics.initImage0.(Ljavax/microedition/lcdui/Image;II)V"] = function(img, w, h) {
-        this.context2D = img.imageData.context;
-        this.context2D.save();
-        this.img = img;
-        setDimensions(this, w, h);
-        resetGC(this);
-    };
+    GraphicsInfo.prototype.translate = function(x, y) {
+        x = x | 0;
+        y = y | 0;
+        if (x !== 0 && y !== 0) {
+            this.transX += x;
+            this.transY += y;
 
-    function isScreenGraphics(g) {
-        return g.displayId != -1;
+            if (this.contextInfo.currentlyAppliedGraphicsInfo === this) {
+                this.contextInfo.currentlyAppliedGraphicsInfo = null;
+            }
+        }
     }
 
-    function resetGC(g) {
-        g.currentFont = getDefaultFont();
-        g.style       = SOLID;
-        g.rgbColor    = g.gray = 0;
-        g.pixel       = getPixel(g.rgbColor, g.gray, true);
-    }
+    GraphicsInfo.prototype.setClip = function(x, y, width, height, minX, minY, maxX, maxY) {
+        var newX1 = x + this.transX;
+        var newY1 = y + this.transY;
+        var newX2 = newX1 + width;
+        var newY2 = newY1 + height;
+
+        newX1 = Math.max(minX, newX1) & 0x7fff;
+        newY1 = Math.max(minY, newY1) & 0x7fff;
+        newX2 = Math.min(maxX, newX2) & 0x7fff;
+        newY2 = Math.min(maxY, newY2) & 0x7fff;
+
+        if (width <= 0 || height <= 0 || newX2 <= newX1 || newY2 <= newY1) {
+            newX1 = newY1 = newX2 = newY2 = 0;
+        }
 
-    function reset(g, x1, y1, x2, y2) {
-        resetGC(g);
-        translate(g, -g.transX, -g.transY);
-        setClip(g, x1, y1, x2 - x1, y2 - y1);
+        if (this.clipX1 === newX1 && this.clipY1 === newY1 && this.clipX2 === newX2 && this.clipY2 === newY2) {
+            return;
+        }
+
+        if (this.contextInfo.currentlyAppliedGraphicsInfo === this) {
+            this.contextInfo.currentlyAppliedGraphicsInfo = null;
+        }
+
+        this.clipX1 = newX1;
+        this.clipY1 = newY1;
+        this.clipX2 = newX2;
+        this.clipY2 = newY2;
     }
 
-    function translate(g, x, y) {
-        g.transX += x;
-        g.transY += y;
-        g.context2D.translate(x, y);
+    GraphicsInfo.prototype.getGraphicsContext = function() {
+        if (this.contextInfo.currentlyAppliedGraphicsInfo !== this) {
+            this.contextInfo.applyGraphics(this);
+        }
+
+        return this.contextInfo.context;
     }
 
-    function setDimensions(g, w, h) {
-      g.maxWidth = w & 0x7fff;
-      g.maxHeight = h & 0x7fff;
-      translate(g, -g.transX, -g.transY);
-      setClip(g, 0, 0, g.maxWidth, g.maxHeight);
+    function ContextInfo(ctx) {
+        this.currentlyAppliedGraphicsInfo = null;
```

Comment:
Changes to this property will change the shape of the ContextInfo object, although maybe that's irrelevant or necessary in any case.